packages/aggregator/src/analytics/aggregateAnalytics/adjustOptionsToAggregationList.js

```diff
@@ -0,0 +1,134 @@
+};
+
+/**
+ * Currently not all `startDate`, `endDate`, `period` options are provided in each client request.
+ * This function preserves the existing functionality by retaining the original period if
+ * there was no date range adjustment
+ */
+const getAdjustedPeriod = (originalOptions, adjustedOptions) => {
+  const areDateRangesEqual =
+    originalOptions?.startDate === adjustedOptions?.startDate &&
+    originalOptions?.endDate === adjustedOptions?.endDate;
+
+  return areDateRangesEqual
+    ? originalOptions.period
+    : convertDateRangeToPeriodString(adjustedOptions.startDate, adjustedOptions.endDate);
+};
+
+const getAdjustedDateRange = (dateRange, aggregationList) =>
+  aggregationList.reduce((currentRange, aggregation) => {
+    const getDateRange = AGGREGATION_TYPE_TO_DATE_RANGE_GETTER[aggregation.type];
+    return getDateRange ? getDateRange(currentRange, aggregation.config) : currentRange;
+  }, dateRange);
+
+const shouldFetchDataSourceEntities = ({ config: { dataSourceEntityType } = {} } = {}) =>
+  dataSourceEntityType;
+
+const shouldFetchRelations = ({
+  type,
+  config: { aggregationEntityType, dataSourceEntityType } = {},
+} = {}) => aggregationEntityType && dataSourceEntityType && !(type === AGGREGATION_TYPES.RAW);
+
+const getAdjustedOrganisationUnitsAndAggregations = async (
+  context,
+  fetchOptions,
+  aggregationList,
+) => {
+  const { hierarchy, organisationUnitCodes } = fetchOptions;
+  if (!aggregationList.some(shouldFetchDataSourceEntities)) {
+    return [organisationUnitCodes, aggregationList];
+  }
+
+  const { session } = context;
+
+  let adjustedOrganisationUnitCodes;
+  const adjustedAggregations = [];
+  const entityConnection = new EntityConnection(session);
+  await Promise.all(
+    aggregationList.map(async aggregation => {
+      if (!shouldFetchDataSourceEntities(aggregation)) {
+        adjustedAggregations.push(aggregation);
+        return;
+      }
+
+      if (!adjustedOrganisationUnitCodes && !shouldFetchRelations(aggregation)) {
+        const { dataSourceEntityType, dataSourceEntityFilter } = aggregation.config;
+        adjustedOrganisationUnitCodes = await entityConnection.getDataSourceEntities(
+          hierarchy,
+          organisationUnitCodes,
+          dataSourceEntityType,
+          dataSourceEntityFilter,
+        );
+        adjustedAggregations.push(aggregation);
+        return;
+      }
+
+      const {
+        aggregationEntityType,
+        dataSourceEntityType,
+        dataSourceEntityFilter,
+      } = aggregation.config;
+      const [
+        dataSourceEntities,
+        relations,
+      ] = await entityConnection.getDataSourceEntitiesAndRelations(
+        hierarchy,
+        organisationUnitCodes,
+        aggregationEntityType,
+        dataSourceEntityType,
+        dataSourceEntityFilter,
+      );
+
+      if (!adjustedOrganisationUnitCodes) {
+        adjustedOrganisationUnitCodes = dataSourceEntities;
```

Comment:
I see that here (and in line 70 above) `adjustedOrganisationUnitCodes` are set, not appended... Intuitively, I would expect them to be a list of items that can build of with each iteration.

On the other hand, I don't know if this would actually make _sense_: in the same way, I haven't thought whether chaining multiple org unit aggregations makes sense in general, or is even used currently (probably not).

Happy to leave this as is, and not make a decision right now, just something to flag for the future - perhaps we should also add a comment here ("chaining org unit aggregations has unexpected behaviour"). We can also throw an error, in this scenario - or skip this for now, to avoid unnecessary regression risk

Reply:
The whole `dataSourceEntities` thing is a bit of a confusing thing in aggregation imo. Truth be told, it's a fetchOption, not an aggregation option. You're saying, fetch me data against all the entities of this type under the requested entity. We even replace the `orgUnitCodes` in the fetch option rather than the aggregation option!

My preferences for how we ought to design it would be:
a) dataSourceEntityType is a fetchOption, which we use to replace the requested org units with as we do here
b) dataSourceEntityType is a top-level property on the aggregation options, which we use to replace the requested org units with as we do here
c) what we currently do, where it's nested inside an aggregation config, and it only makes sense for us to fetch the entities for the very first aggregation which specifies the dataSourceEntityType...


Btw I actually think chaining entity aggregations makes sense and would work with the implementation I have here, eg:
```
fetchOptions = { organisationUnitCodes: ['TO'], hierarchy: 'explore' }

aggregations = [
{
  type: 'FINAL_PER_ORG_GROUP', 
  config: { dataSourceEntityType: 'village', aggregationEntityType: 'facility'}
}, 
{
  type: 'SUM_PER_ORG_GROUP', 
  config: { dataSourceEntityType: 'facility', aggregationEntityType: 'country'}
}]

// After running adjustOptionsToAggregationList

fetchOptions = { organisationUnitCodes: [<all_villages_in_tonga>], hierarchy: 'explore' }

aggregations = [
{
  type: 'FINAL_PER_ORG_GROUP', 
  config: { dataSourceEntityType: 'village', aggregationEntityType: 'facility', orgUnitMap: <village->facility>}
}, 
{
  type: 'SUM_PER_ORG_GROUP', 
  config: { dataSourceEntityType: 'facility', aggregationEntityType: 'country', orgUnitMap: <facility->country>}
}]
``` 
I believe this should correctly compute the sum of the final villlage per facility per country 😄

Reply:
It should! It seems to depend on the aggregation order, which is a reasonable requirement in my opinion - after all, the client should be aware that aggregations are performed in order.

Great point about `dataSourceEntities` being a fetch option in essence! We may find that we can improve this even further - moving all logic here is a the great, big step! 🚀